Added tbe ML Workflow and the spatial_distribution class

jdiego-miyashiro opened the pull request:
A purely categorical dataset mushroom.csv was added along with an exploration of its content, a paper on how to measure distance and similarity with categorical data was include in the bibliography and finally, the first version of file spatial_distribution.py is included, at the moment it only can serve to analyze data that is exclusively categorical.

dzeber commented on datasets/Exploration of mushroom dataset/Mushrooms_edibility.ipynb:
```diff
@@ -0,0 +1,530 @@
+    "import seaborn as sns\n",
+    "import matplotlib.pyplot as plt\n",
+    "from sklearn.compose import ColumnTransformer\n",
+    "from sklearn.preprocessing import Imputer, LabelEncoder, OneHotEncoder\n",
```
Edit: in fact, it doesn't look like this is used, so maybe just drop it.

jdiego-miyashiro replied:
Done

dzeber reviewed (changes requested):
This is good work! I like this dataset, nice to have a decent-sized, purely categorical example to work with.

There are a couple of organizational changes I'd like to make:
- Please move the notebook to `examples/data_exploration/`. We will maintain the ML workflow notebooks there.
- Does your code require the `new_data` auxiliary file? If not, let's drop it from the PR
- Please resolve the conflict in the datasets README (this happened because we merged another PR that modified this file) - feel free to ping me if you need help with this.
- Please break out your work on the spatial distributions to a separate PR. That way we can keep this one for the ML workflow notebook and dataset, and discuss your package code over there.

The notebook is clear and easy to follow. I've added a couple of specific comments inline in the code itself. A couple of more general comments I had:

- In general, it's a good idea to sprinkle markdown cells throughout your notebook to describe your thought process, findings, interpretation, etc. I see you've added these as code comments, but it would be good to move them to separate markdown cells above the code blocks.
- It looks like you've used 5% training in the train-test split? That's an unusually small training set, even though it seems to work well. What guided your choice here?
- It seems like the logistic classifier is really good at discriminating between the classes - that might be why you only needed a small training set. Any intuition into why this is the case? Are the classes just really well separated?

Great start on the `spatial_distribution` code. I will follow up with more comments in the the new PR. Please make sure to include tests in a module under `tests/`, and it's a good idea to also add a notebook under `examples` showing the output when applied to your dataset.

jdiego-miyashiro commented:
Hi David,

new_data.csv was meant to be removed I had it there at the beginning to
debug some issues, and yes I think ill most likely need help to merge the
readme I try to do it yesterday in my own fork and failed. I ended up using
5% because that indeed gave me misclassified points greater percentage
ended up classifying the data almost perfectly. I'll be moving everything
as indicated. Also, I've been investigating a bit more what can we do with
the distance metrics that we have now and I found this very
interesting video on clustering https://www.youtube.com/watch?v=5cOhL4B5waU
I was wondering if it's something you might be interested in having
implemented.

Juan Diego



On Fri, Oct 23, 2020 at 6:13 PM David Zeber <notifications@github.com>
wrote:

> *@dzeber* requested changes on this pull request.
>
> This is good work! I like this dataset, nice to have a decent-sized,
> purely categorical example to work with.
>
> There are a couple of organizational changes I'd like to make:
>
>    - Please move the notebook to examples/data_exploration/. We will
>    maintain the ML workflow notebooks there.
>    - Does your code require the new_data auxiliary file? If not, let's
>    drop it from the PR
>    - Please resolve the conflict in the datasets README (this happened
>    because we merged another PR that modified this file) - feel free to ping
>    me if you need help with this.
>    - Please break out your work on the spatial distributions to a
>    separate PR. That way we can keep this one for the ML workflow notebook and
>    dataset, and discuss your package code over there.
>
> The notebook is clear and easy to follow. I've added a couple of specific
> comments inline in the code itself. A couple of more general comments I had:
>
>    - In general, it's a good idea to sprinkle markdown cells throughout
>    your notebook to describe your thought process, findings, interpretation,
>    etc. I see you've added these as code comments, but it would be good to
>    move them to separate markdown cells above the code blocks.
>    - It looks like you've used 5% training in the train-test split?
>    That's an unusually small training set, even though it seems to work well.
>    What guided your choice here?
>    - It seems like the logistic classifier is really good at
>    discriminating between the classes - that might be why you only needed a
>    small training set. Any intuition into why this is the case? Are the
>    classes just really well separated?
>
> Great start on the spatial_distribution code. I will follow up with more
> comments in the the new PR. Please make sure to include tests in a module
> under tests/, and it's a good idea to also add a notebook under examples
> showing the output when applied to your dataset.
> ------------------------------
>
> In datasets/Exploration of mushroom dataset/Mushrooms_edibility.ipynb
> <https://github.com/mozilla/PRESC/pull/225#discussion_r511201172>:
>
> > @@ -0,0 +1,530 @@
> +{
> + "cells": [
> +  {
> +   "cell_type": "code",
> +   "execution_count": 13,
> +   "metadata": {},
> +   "outputs": [],
> +   "source": [
> +    "import pandas as pd\n",
> +    "import numpy as np\n",
> +    "import seaborn as sns\n",
> +    "import matplotlib.pyplot as plt\n",
> +    "from sklearn.compose import ColumnTransformer\n",
> +    "from sklearn.preprocessing import Imputer, LabelEncoder, OneHotEncoder\n",
>
> the Imputer import has changed to sklearn.impute.SimpleImputer in our
> version of scikit-learn
> ------------------------------
>
> In datasets/Exploration of mushroom dataset/Mushrooms_edibility.ipynb
> <https://github.com/mozilla/PRESC/pull/225#discussion_r511201269>:
>
> > +       "1           o         p                 n          n       g  \n",
> +       "2           o         p                 n          n       m  \n",
> +       "3           o         p                 k          s       u  \n",
> +       "4           o         e                 n          a       g  \n",
> +       "\n",
> +       "[5 rows x 23 columns]"
> +      ]
> +     },
> +     "execution_count": 14,
> +     "metadata": {},
> +     "output_type": "execute_result"
> +    }
> +   ],
> +   "source": [
> +    "#load the data and display the first 5 rows\n",
> +    "rooms=pd.read_csv('mushrooms.csv')\n",
>
> This path needs to work relative to the dataset location in the repo
> ------------------------------
>
> In datasets/Exploration of mushroom dataset/Mushrooms_edibility.ipynb
> <https://github.com/mozilla/PRESC/pull/225#discussion_r511201483>:
>
> > +   "execution_count": 23,
> +   "metadata": {},
> +   "outputs": [],
> +   "source": [
> +    "#split data\n",
> +    "x_train, x_test, y_train, y_test = train_test_split(X, y, train_size = 0.05,random_state=0)"
> +   ]
> +  },
> +  {
> +   "cell_type": "code",
> +   "execution_count": 25,
> +   "metadata": {},
> +   "outputs": [],
> +   "source": [
> +    "#train our model\n",
> +    "classifier.fit(x_train,y_train)\n",
>
> The definition of classifier needs to precede this (I saw it farther down)
> ------------------------------
>
> In datasets/Exploration of mushroom dataset/Mushrooms_edibility.ipynb
> <https://github.com/mozilla/PRESC/pull/225#discussion_r511203912>:
>
> > @@ -0,0 +1,530 @@
> +{
> + "cells": [
> +  {
> +   "cell_type": "code",
> +   "execution_count": 13,
> +   "metadata": {},
> +   "outputs": [],
> +   "source": [
> +    "import pandas as pd\n",
> +    "import numpy as np\n",
> +    "import seaborn as sns\n",
> +    "import matplotlib.pyplot as plt\n",
> +    "from sklearn.compose import ColumnTransformer\n",
> +    "from sklearn.preprocessing import Imputer, LabelEncoder, OneHotEncoder\n",
>
> Edit: in fact, it doesn't look like this is used, so maybe just drop it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/PRESC/pull/225#pullrequestreview-516123575>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANVGZBJXOE52ZH2PMHDE5ETSMILZFANCNFSM4SUSAETQ>
> .
>

dzeber commented:
Let's discuss the merge conflict thing in the meeting tomorrow.

Glad you're looking into further approaches like clustering! That could be useful here - however I think we should wait until we've tried out the current approach on some example datasets to see what the shortcomings are.

jdiego-miyashiro commented:
Hi David,

I added the plot comparison between the average distance from the misclassified points to every other point vs the distance from every other regular point to the other regular points. Also added one more similarity measure from the paper called lin distance

dzeber reviewed (changes requested):
Nice to have these distance metrics implemented. I'd now like to focus on getting this cleaned up so we can merge.

- Please address the outstanding comments for the notebook and in the module code.
- Please move the  `graph_examples` folder inside the `examples` dir.

You also need to add a proper test suite. This should be a module inside [`tests/`](https://github.com/mozilla/PRESC/tree/master/tests) with a function named `test_<...>` for each of the main functions in your categorical_distance module. Eg. for the metrics, you should test them on a small toy dataset just to make sure the logic works as expected. See the [pytest doc](https://docs.pytest.org/en/latest/index.html) for more info. As we discussed with the group, to test the plotting functions, just call the function and make sure it runs without error.

As far as I can tell, `categorical_distance` still doesn't offer the functionality described in the [roadmap](https://github.com/mozilla/PRESC/blob/master/docs/ROADMAP.md): histograms of distances from a specific point to the other misclassified and correctly classified points, split by class. It looks like the groundwork is there, but you should provide a function that runs that analysis as well.

What are your thoughts on choosing between the different metrics you have implemented, ie  are some more appropriate in certain situations? If you have findings on this, it would be good to document.

The next step would be to handle continuous data. Let's get this PR merged first and then you can work on that next.

jdiego-miyashiro commented on datasets/Exploration of mushroom dataset/Mushrooms_edibility.ipynb:
```diff
@@ -0,0 +1,530 @@
+   ],
+   "source": [
+    "#load the data and display the first 5 rows\n",
+    "rooms=pd.read_csv('mushrooms.csv')\n",
```
Done

jdiego-miyashiro commented on datasets/Exploration of mushroom dataset/Mushrooms_edibility.ipynb:
```diff
@@ -0,0 +1,530 @@
+   "outputs": [],
+   "source": [
+    "#train our model\n",
+    "classifier.fit(x_train,y_train)\n",
```
Done

jdiego-miyashiro commented on examples/Mushrooms_edibility.ipynb:
```diff
@@ -0,0 +1,1006 @@
+    "import seaborn as sns\n",
+    "import matplotlib.pyplot as plt\n",
+    "from sklearn.compose import ColumnTransformer\n",
+    "from sklearn.preprocessing import Imputer, LabelEncoder, OneHotEncoder\n",
```
Done

jdiego-miyashiro commented on examples/Mushrooms_edibility.ipynb:
```diff
@@ -0,0 +1,1006 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import categorical_distance as catd\n",
```
Done

jdiego-miyashiro commented on presc/spatial_distribution/categorical_distance.py:
```diff
@@ -0,0 +1,509 @@
+        """Computes the overlap similarity measure for categorical data"""
+
+        if list(dpoint1.index) == list(dpoint2.index):
+            overlap_score = len(set(dpoint1) & set(dpoint2)) / len(dpoint1)
```
Done

jdiego-miyashiro commented on presc/spatial_distribution/categorical_distance.py:
```diff
@@ -0,0 +1,509 @@
+
+                for count in counts:
+                    attribute_score = self.goodall_frequency(count) + attribute_score
+                    pass
```
Removed

jdiego-miyashiro commented on presc/spatial_distribution/categorical_distance.py:
```diff
@@ -0,0 +1,509 @@
+            self.__buildcounts()
+        )  # dictionary of counts of occurances of attribute instances
+
+        self._KNNmodel = None
```
At some point, I was considering passing a KNN model from sklearn to implement things like finding the closest k neighbors, k-means, etc. I ended up leaving this as a placeholder

jdiego-miyashiro commented on presc/spatial_distribution/categorical_distance.py:
```diff
@@ -0,0 +1,509 @@
+
+        return distance_array
+
+    def plot_knearest_points(self, dpoint, metric1, metric2, metric3, k):
```
Done

jdiego-miyashiro commented on presc/spatial_distribution/categorical_distance.py:
```diff
@@ -0,0 +1,509 @@
+        plt.show()
+        return
+
+    def plot_distance_misclasified(
```
Done

jdiego-miyashiro commented on presc/spatial_distribution/categorical_distance.py:
```diff
@@ -0,0 +1,509 @@
+
+    def goodall2(self, dpoint1, dpoint2):
+        """Computes the goodall2 similary measurement for categorical data, see paper by
+        Varun, Shyam and Vipin in the bibliography carpet for reference"""
```
Yes you are right I missed this

jdiego-miyashiro commented on presc/spatial_distribution/categorical_distance.py:
```diff
@@ -0,0 +1,509 @@
+        d1_attributes = list(dpoint1.index)
+        d2_attributes = list(dpoint2.index)
+
+        if d1_attributes == d2_attributes:
```
Removed

jdiego-miyashiro commented on examples/Mushrooms_edibility.ipynb:
```diff
@@ -0,0 +1,1006 @@
+   ],
+   "source": [
+    "#load the data and display the first 5 rows\n",
+    "rooms=pd.read_csv('C:/Users/castromi/Documents/GitHub/PRESC/datasets/mushrooms.csv')\n",
```
Done

jdiego-miyashiro commented on presc/spatial_distribution/categorical_distance.py:
```diff
@@ -0,0 +1,509 @@
+                attribute_score = 0
+                counts = self._counts_per_attribute[attribute]
+
+                for count in counts:
```
yes you are right I completely missed this

jdiego-miyashiro commented on presc/spatial_distribution/categorical_distance.py:
```diff
@@ -0,0 +1,509 @@
+
+        return 1 / weight_denominator
+
+        return
```
done

jdiego-miyashiro commented on presc/spatial_distribution/categorical_distance.py:
```diff
@@ -0,0 +1,509 @@
+        return weight * lin
+
+    def lin_weight(self, dpoint1, dpoint2):
+        d1_attributes = list(dpoint1.index)
```
I wanted to keep it separate because I thought it cluttered the code and having it separately was more neat. I can integrate it if you think that's easier to follow

dzeber replied:
I agree in general breaking computations into smaller functions is a good practice, but here it seems unnecessary to recompute all the empirical frequencies. I would just accumulate the `weight_denominator` inside the `for` loop in `lin()`.

jdiego-miyashiro commented on presc/spatial_distribution/categorical_distance.py:
```diff
@@ -0,0 +1,509 @@
+
+        if distance_sample < 1:
+            distance_sample = round(distance_sample * self._data_len)
+        elif distance_sample > self._data_len:
```
the user can input a number of sample points the function can use. I originally forgot to add this to the documentation. In case the number is greater than 1 the program will use it as the exact number of points used to estimate the average distance

jdiego-miyashiro commented on presc/spatial_distribution/categorical_distance.py:
```diff
@@ -0,0 +1,509 @@
+                             by the model
+            label_true: List or pandas series that holds the true labels
+            model:Currently not implemented and set to None, in the future it might ask for the specific
+                    sklearn trained model for more functionalities
```
added

dzeber reviewed (commented):
Thanks for making these updates. Please drop the changes to the notebooks other than the mushroom notebook from this PR. Once you have done that, we can merge this, but as a follow-up I would like you to prepare a new PR adding what was requested in the previous comment: a test suite and a function providing the functionality described in the roadmap.

jdiego-miyashiro commented:
Hi David, I'm confused I don't think I made any changes in any other
notebooks that were not the one I was working on. Also yeah I was already
working on the test suit, the methods plot_distance_histogram and
plot_full_histogram_report are already implementing the functionality
described in the roadmap. I updated the mushroom notebook so it's shown
there as part of the exploration of the dataset

On Mon, Nov 16, 2020 at 3:14 PM David Zeber <notifications@github.com>
wrote:

> *@dzeber* commented on this pull request.
>
> Thanks for making these updates. Please drop the changes to the notebooks
> other than the mushroom notebook from this PR. Once you have done that, we
> can merge this, but as a follow-up I would like you to prepare a new PR
> adding what was requested in the previous comment: a test suite and a
> function providing the functionality described in the roadmap.
> ------------------------------
>
> In presc/spatial_distribution/categorical_distance.py
> <https://github.com/mozilla/PRESC/pull/225#discussion_r524657537>:
>
> > +            weight = self.lin_weight(dpoint1, dpoint2)
> +            for attribute in d1_attributes:
> +                frequency_dpoint1 = self.empirical_frequency(
> +                    dpoint1[attribute], attribute
> +                )
> +                frequency_dpoint2 = self.empirical_frequency(
> +                    dpoint2[attribute], attribute
> +                )
> +                if dpoint1[attribute] == dpoint2[attribute]:
> +                    lin = lin + 2 * np.log(frequency_dpoint1)
> +                else:
> +                    lin = lin + 2 * np.log(frequency_dpoint1 + frequency_dpoint2)
> +        return weight * lin
> +
> +    def lin_weight(self, dpoint1, dpoint2):
> +        d1_attributes = list(dpoint1.index)
>
> I agree in general breaking computations into smaller functions is a good
> practice, but here it seems unnecessary to recompute all the empirical
> frequencies. I would just accumulate the weight_denominator inside the for
> loop in lin().
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/PRESC/pull/225#pullrequestreview-531827123>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANVGZBIUNLKCKC23IUI7SBLSQGP25ANCNFSM4SUSAETQ>
> .
>

jdiego-miyashiro commented:
Hi nevermind I found the changes I made and reverted them

On Mon, Nov 16, 2020 at 4:36 PM Juan Diego Castro Miyashiro <
castromi@ualberta.ca> wrote:

> Hi David, I'm confused I don't think I made any changes in any other
> notebooks that were not the one I was working on. Also yeah I was already
> working on the test suit, the methods plot_distance_histogram and
> plot_full_histogram_report are already implementing the functionality
> described in the roadmap. I updated the mushroom notebook so it's shown
> there as part of the exploration of the dataset
>
> On Mon, Nov 16, 2020 at 3:14 PM David Zeber <notifications@github.com>
> wrote:
>
>> *@dzeber* commented on this pull request.
>>
>> Thanks for making these updates. Please drop the changes to the notebooks
>> other than the mushroom notebook from this PR. Once you have done that, we
>> can merge this, but as a follow-up I would like you to prepare a new PR
>> adding what was requested in the previous comment: a test suite and a
>> function providing the functionality described in the roadmap.
>> ------------------------------
>>
>> In presc/spatial_distribution/categorical_distance.py
>> <https://github.com/mozilla/PRESC/pull/225#discussion_r524657537>:
>>
>> > +            weight = self.lin_weight(dpoint1, dpoint2)
>> +            for attribute in d1_attributes:
>> +                frequency_dpoint1 = self.empirical_frequency(
>> +                    dpoint1[attribute], attribute
>> +                )
>> +                frequency_dpoint2 = self.empirical_frequency(
>> +                    dpoint2[attribute], attribute
>> +                )
>> +                if dpoint1[attribute] == dpoint2[attribute]:
>> +                    lin = lin + 2 * np.log(frequency_dpoint1)
>> +                else:
>> +                    lin = lin + 2 * np.log(frequency_dpoint1 + frequency_dpoint2)
>> +        return weight * lin
>> +
>> +    def lin_weight(self, dpoint1, dpoint2):
>> +        d1_attributes = list(dpoint1.index)
>>
>> I agree in general breaking computations into smaller functions is a good
>> practice, but here it seems unnecessary to recompute all the empirical
>> frequencies. I would just accumulate the weight_denominator inside the
>> for loop in lin().
>>
>> —
>> You are receiving this because you modified the open/close state.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/mozilla/PRESC/pull/225#pullrequestreview-531827123>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ANVGZBIUNLKCKC23IUI7SBLSQGP25ANCNFSM4SUSAETQ>
>> .
>>
>

dzeber commented:
The roadmap functionality would be the distribution (eg. histogram) of the distances between a specific point and all the other points, split by class. This is similar to `distance_to_data`, except it would be the distribution of distances rather than the average distance. As far as I can tell, the `plot_histogram` methods show the distributions of average distances acrosss all points.

I'm going to merge this as is, and you can follow up with the tests in a new PR.

jdiego-miyashiro commented:
Oh okay I see, I will implement it right away, it shouldn't be too hard
since I can reuse the same machinery as histo_sample setting the sample
size we use to estimate the average distance across all points to one point
and forcing it to be in particular that single specified point

On Mon, Nov 16, 2020 at 6:14 PM David Zeber <notifications@github.com>
wrote:

> The roadmap functionality would be the distribution (eg. histogram) of the
> distances between a specific point and all the other points, split by
> class. This is similar to distance_to_data, except it would be the
> distribution of distances rather than the average distance. As far as I can
> tell, the plot_histogram methods show the distributions of average
> distances acrosss all points.
>
> I'm going to merge this as is, and you can follow up with the tests in a
> new PR.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/PRESC/pull/225#issuecomment-728600302>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANVGZBIS3OY6PQQSEUPVC7TSQHE7JANCNFSM4SUSAETQ>
> .
>
